Sources/Kitura/CodableRouter+TypedMiddleware.swift

```diff
@@ -0,0 +1,1020 @@
+    }
+    
+    public func put<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable>(_ route: String, handler: @escaping TwoMiddlewareIdentifierCodableClosure<T1, T2, Id, I, O>) {
+        putSafely(route, handler: handler)
+    }
+    
+    public func put<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, T3: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable>(_ route: String, handler: @escaping ThreeMiddlewareIdentifierCodableClosure<T1, T2, T3, Id, I, O>) {
+        putSafely(route, handler: handler)
+    }
+    
+    // PATCH with identifier
+    public func patch<T: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable>(_ route: String, handler: @escaping MiddlewareIdentifierCodableClosure<T, Id, I, O>) {
+        patchSafely(route, handler: handler)
+    }
+    
+    public func patch<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable>(_ route: String, handler: @escaping TwoMiddlewareIdentifierCodableClosure<T1, T2, Id, I, O>) {
+        patchSafely(route, handler: handler)
+    }
+    
+    public func patch<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, T3: TypeSafeMiddleware, Id: Identifier, I: Codable, O: Codable>(_ route: String, handler: @escaping ThreeMiddlewareIdentifierCodableClosure<T1, T2, T3, Id, I, O>) {
+        patchSafely(route, handler: handler)
+    }
+    
+    // POST using one middleware
+    fileprivate func postSafely<T: TypeSafeMiddleware, I: Codable, O: Codable>(_ route: String, handler: @escaping MiddlewareCodableClosure<T, I, O>) {
+        post(route) { request, response, next in
+            Log.verbose("Received POST type-safe request")
+            guard let codableInput = CodableHelpers.readCodableOrSetResponseStatus(I.self, from: request, response: response) else {
+                next()
+                return
+            }
+            self.handleMiddleware(T.self, request: request, response: response) { TypeSafeMiddleware in
+                guard let TypeSafeMiddleware = TypeSafeMiddleware else {
+                    return next()
+                }
+                handler(TypeSafeMiddleware, codableInput, CodableHelpers.constructOutResultHandler(successStatus: .created, response: response, completion: next))
+            }
+        }
+    }
+    
+    // POST using two middleware
+    fileprivate func postSafely<T1: TypeSafeMiddleware, T2: TypeSafeMiddleware, I: Codable, O: Codable>(_ route: String, handler: @escaping TwoMiddlewareCodableClosure<T1, T2, I, O>) {
+        post(route) { request, response, next in
+            Log.verbose("Received POST type-safe request")
+            guard let codableInput = CodableHelpers.readCodableOrSetResponseStatus(I.self, from: request, response: response) else {
+                next()
+                return
+            }
+            self.handleMiddleware(T1.self, T2.self, request: request, response: response) { TypeSafeMiddleware1, TypeSafeMiddleware2 in
```

Comment:
Please name variables with lower-case names, e.g. `typeSafeMiddleware1` not `TypeSafeMiddleware1` to distinguish them from type names.  This comment applies across the whole PR.